bundles/org.openhab.core.model.yaml/src/main/java/org/openhab/core/model/yaml/internal/rules/YamlRuleTemplateProvider.java

```diff
@@ -0,0 +1,172 @@
+import org.osgi.service.component.annotations.Deactivate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link YamlRuleTemplateProvider} is an OSGi service, that allows definition of rule templates in YAML configuration
+ * files. Files can
+ * be added, updated or removed at runtime. The rule templates are automatically registered with
+ * {@link org.openhab.core.automation.internal.template.RuleTemplateRegistry}.
+ *
+ * @author Ravi Nadahar - Initial contribution
+ */
+@NonNullByDefault
+@Component(immediate = true, service = { RuleTemplateProvider.class, YamlRuleTemplateProvider.class,
+        YamlModelListener.class })
+public class YamlRuleTemplateProvider extends AbstractYamlRuleProvider<RuleTemplate>
+        implements RuleTemplateProvider, YamlModelListener<YamlRuleTemplateDTO> {
+
+    private final Logger logger = LoggerFactory.getLogger(YamlRuleTemplateProvider.class);
+
+    private final Map<String, Collection<RuleTemplate>> ruleTemplatesMap = new ConcurrentHashMap<>();
+
+    @Activate
+    public YamlRuleTemplateProvider() {
+    }
+
+    @Deactivate
+    public void deactivate() {
+        ruleTemplatesMap.clear();
+    }
+
+    @Override
+    public Collection<RuleTemplate> getAll() {
+        return getTemplates(null);
+    }
+
+    @Override
+    public @Nullable RuleTemplate getTemplate(String uid, @Nullable Locale locale) {
+        return ruleTemplatesMap.values().stream().flatMap(list -> list.stream()).filter(t -> uid.equals(t.getUID()))
+                .findAny().orElse(null);
+    }
+
+    @Override
+    public Collection<RuleTemplate> getTemplates(@Nullable Locale locale) {
+        return ruleTemplatesMap.values().stream().flatMap(list -> list.stream()).toList();
+    }
+
+    @Override
+    public Class<YamlRuleTemplateDTO> getElementClass() {
+        return YamlRuleTemplateDTO.class;
+    }
+
+    @Override
+    public boolean isVersionSupported(int version) {
+        return version >= 1;
+    }
+
+    @Override
+    public boolean isDeprecated() {
+        return false;
+    }
+
+    @Override
+    public void addedModel(String modelName, Collection<YamlRuleTemplateDTO> elements) {
+        List<RuleTemplate> added = elements.stream().map(this::mapRuleTemplate).filter(Objects::nonNull).toList();
+        Collection<RuleTemplate> modelRuleTemplates = Objects
+                .requireNonNull(ruleTemplatesMap.computeIfAbsent(modelName, k -> new ArrayList<>()));
+        modelRuleTemplates.addAll(added);
+        added.forEach(t -> {
+            logger.debug("model {} added rule template {}", modelName, t.getUID());
+            notifyListenersAboutAddedElement(t);
+        });
+    }
+
+    @Override
+    public void updatedModel(String modelName, Collection<YamlRuleTemplateDTO> elements) {
+        List<RuleTemplate> updated = elements.stream().map(this::mapRuleTemplate).filter(Objects::nonNull).toList();
+        Collection<RuleTemplate> modelRules = Objects
```

Comment:
modelRules to be renamed into modelRuleTemplates

Reply:
done